### DISCUSSION TITLE
A Toga interactive plotting facility?

### DISCUSSION THREAD

**Discussion by paolofranzetti** [beeware/toga#2813] [Ideas]
I am a python software developer working to astronomical data analysis software.

I recently discovered Beeware and Toga and I think they could be extremely useful for the astronomical (and, in general, scientific) community. 

However what is really missing is an interactive plotting library. Interactive plots are widely used during scientific analysis and this kind of widget could give a boost to Toga usage in this context.

I think that starting from the toga.Canvas widget it could be quite easy to develop a plotting Toga module which implements an API similar to the matplotlib one but with the addiction of interactivity. 

As a test I even wrote a brief script that plots a galactic spectrum and zooms into it using the mouse; just a few lines of code.

So here is the big question; are the "toga managers" interested in this idea ?

**Comment by freakboy3742** [comment]
Are we interested in the general concept? Sure! It sounds like an interesting use case for Toga; and I agree that science visualisation could be a really good use case for Toga.

However, does this feature need to be part of Toga itself? Almost certainly not. A charting library is the sort of feature that could happily live as a third-party library, imported by those projects that need it, using the basic widgets provided by Toga.

As evidence of this - we've already got [toga-chart](https://github.com/beeware/toga-chart). That's a matplotlib backend, backed by Toga's canvas widget. 

It doesn't *currently* implement of the interactive features of matplotlib, but that's entirely because we haven't been focussing on that widget specifically. If someone wanted to add support for interactive charts to toga-chart, I'd gladly look at any contributions; and if supporting interactivity in charts required additional to the base functionality of Canvas, I'd be happy to consider those changes as well.

**Reply by paolofranzetti** [reply]
Thanks for your answer and your interest. I saw the toga-chart module, but I must admit that I do not like very much the matplotlib look-and-feel (not to mention its "publication quality" plots) and its interactivity features. So I would prefer something more "native" on which I have full control.

But in Italy we say "tra il dire e il fare c'è di mezzo il mare" which literally translates into "from saying to doing there is the sea" ;-)

I will keep working on my plotting module; if and when it will became something I am not too ashamed of, I will propose it to you (if you are still interested at that time).

About your "supporting" offer, the only thing I currently miss in the Canvas widget is an handler linked to mouse wheel click and scrolling. Is it foreseen? Can I submit an "official" request for it?

**Reply by freakboy3742** [reply]
> Thanks for your answer and your interest. I saw the toga-chart module, but I must admit that I do not like very much the matplotlib look-and-feel (not to mention its "publication quality" plots) and its interactivity features. So I would prefer something more "native" on which I have full control.

Totally understandable. I won't defend the look and feel of matplotlib - we only implemented that API because it's ubiquitous, and there was a relatively straightforward API to implement to provide support.

> I will keep working on my plotting module; if and when it will became something I am not too ashamed of, I will propose it to you (if you are still interested at that time).

To be clear here - I'm suggesting that the right place for this is *not* part of Toga itself, or even part of BeeWare, but as a third-party package on PyPI that you maintain and others use. A package doesn't need to be "official" to be useful - witness the wide range of packages on PyPI. They're not "official" Python, but that doesn't make them any less useful.

We're happy to provide links to your package if/when the time comes so that others can find it easily. The only request I'd have is that if you want to include "toga" in the name of your package, you use the `togax-` prefix, rather than `toga-`. Toga is a non-registered trademark, and we intend to preserve the `toga-` prefix for *official* packages (like `toga-chart`).
 
> About your "supporting" offer, the only thing I currently miss in the Canvas widget is an handler linked to mouse wheel click and scrolling. Is it foreseen? Can I submit an "official" request for it?

You can certainly submit a request for the feature - the ticket tracker is open for all. However, that doesn't mean it will be implemented immediately by one of the core team. If you want an API to be implemented, the best approach is to submit a pull request.

That said - Toga is a cross platform framework, and the APIs that are implemented need to keep that in mind. The action handler for a button is "press", not "click" - because if you're on a phone, you don't "click" anything. So - if you want to support mouse wheel scrolling, you'll need to propose an abstraction that makes sense on platforms that don't have a mouse wheel - like phones, and laptops with a trackpad.

**Reply by paolofranzetti** [reply]
> To be clear here - I'm suggesting that the right place for this is _not_ part of Toga itself, or even part of BeeWare, but as a third-party package on PyPI that you maintain and others use. A package doesn't need to be "official" to be useful - witness the wide range of packages on PyPI. They're not "official" Python, but that doesn't make them any less useful.
> 
> We're happy to provide links to your package if/when the time comes so that others can find it easily. The only request I'd have is that if you want to include "toga" in the name of your package, you use the `togax-` prefix, rather than `toga-`. Toga is a non-registered trademark, and we intend to preserve the `toga-` prefix for _official_ packages (like `toga-chart`).

I perfectly understood your point and I agree. My idea of "proposing the module to you" was just to let you evaluate it and provide links to my package if you think it's worth it. Ok for the `togax-` prefix.

> You can certainly submit a request for the feature - the ticket tracker is open for all. However, that doesn't mean it will be implemented immediately by one of the core team. 

Sure, I will adapt my development to your timeline, not the reverse

> If you want an API to be implemented, the best approach is to submit a pull request.

Ok, I will submit the request

> That said - Toga is a cross platform framework, and the APIs that are implemented need to keep that in mind. The action handler for a button is "press", not "click" - because if you're on a phone, you don't "click" anything. So - if you want to support mouse wheel scrolling, you'll need to propose an abstraction that makes sense on platforms that don't have a mouse wheel - like phones, and laptops with a trackpad.

A part from the naming, I thought to propose this useful feature (at least for me) because I saw the existence of the `on_alt_*` ones; also these handlers do not make sense on phones (as it is stated by "This event is not supported on Android or iOS."). Conversely, on laptops with a trackpad scrolling is a standard gesture.

**Reply by freakboy3742** [reply]
> > That said - Toga is a cross platform framework, and the APIs that are implemented need to keep that in mind. The action handler for a button is "press", not "click" - because if you're on a phone, you don't "click" anything. So - if you want to support mouse wheel scrolling, you'll need to propose an abstraction that makes sense on platforms that don't have a mouse wheel - like phones, and laptops with a trackpad.
 
> A part from the naming, I thought to propose this useful feature (at least for me) because I saw the existence of the `on_alt_*` ones; also these handlers do not make sense on phones (as it is stated by "This event is not supported on Android or iOS."). Conversely, on laptops with a trackpad scrolling is a standard gesture.

There's a difference between "not implemented" and "can't be implemented". The intention for `on_alt_` is that we'd likely do something like "long press" as the "alt" press on mobile. We haven't *implemented* that feature, but we know how it *would* be implemented, and the terminology *doesn't* require "right_click" (which betrays a specific, right-handed mouse implementation).

Scrolling *is* a standard gesture on trackpads - but we need to be clear about how "normal" UI scrolling interacts with gestures. If you put a canvas inside a scroll window... where does the "scroll" gesture go?

**Reply by paolofranzetti** [reply]
> There's a difference between "not implemented" and "can't be implemented". The intention for `on_alt_` is that we'd likely do something like "long press" as the "alt" press on mobile. We haven't _implemented_ that feature, but we know how it _would_ be implemented, and the terminology _doesn't_ require "right_click" (which betrays a specific, right-handed mouse implementation).

My fault, I did not understood that "not supported" stands for "not implemented"

> Scrolling _is_ a standard gesture on trackpads - but we need to be clear about how "normal" UI scrolling interacts with gestures. If you put a canvas inside a scroll window... where does the "scroll" gesture go?

I think that in the case you described the "scroll" gesture should go in the canvas when the mouse in inside it and in the window otherwise.  

Anyway I do not want to further discuss you development choices; the project is yours, the decisions are yours. 

Just allow a kind advice from an old developer (my "first" python was 1.6). It is great that your framework is natively cross platform; but limiting its functionalities for the only reason to be "fully" cross platform may be counterproductive for its wide diffusion; many science groups develop python GUI applications that no one will ever use on a phone or a tablet.

**Reply by freakboy3742** [reply]
> > Scrolling _is_ a standard gesture on trackpads - but we need to be clear about how "normal" UI scrolling interacts with gestures. If you put a canvas inside a scroll window... where does the "scroll" gesture go?
> 
> I think that in the case you described the "scroll" gesture should go in the canvas when the mouse in inside it and in the window otherwise.

I'm not sure I see how that works. How do you scroll the canvas itself in this case without literally reimplementing the scrolling behavior that ScrollContainer provides (including, but not limited to, actual scroll bars)?

> Just allow a kind advice from an old developer (my "first" python was 1.6). 

Mine too. Don't let my username fool you - I'm a nearing-greybeard who has been at this for a while :-)

> It is great that your framework is natively cross platform; but limiting its functionalities for the only reason to be "fully" cross platform may be counterproductive for its wide diffusion; many science groups develop python GUI applications that no one will ever use on a phone or a tablet.

Sure - and there are features that we explicitly *don't* support on mobile for this reason. However, if you don't keep issues like "how will we implement this on mobile" in mind, you end up painting yourself into corner. I'm totally willing to accept "This isn't supported on mobile" as an answer - *if* there's a reasonable reason why that is the answer.

**Reply by paolofranzetti** [reply]
> > I think that in the case you described the "scroll" gesture should go in the canvas when the mouse in inside it and in the window otherwise.
> 
> I'm not sure I see how that works. How do you scroll the canvas itself in this case without literally reimplementing the scrolling behavior that ScrollContainer provides (including, but not limited to, actual scroll bars)?

I agree that I do not want to reimplement the scrolling behavior of the ScrollContainer. Here I am thinking of use cases when you want to link the scroll event to a different action. An example is a canvas containing a plot where you want to use the mouse wheel to zoom in and out the plot itself; this is a "quite standard" and appreciated behavior of many, if not all, scientific plotting facilities.

**Reply by freakboy3742** [reply]
I don't disagree that it is "quite standard". What I'm highlighting is that there *is* an interaction with existing scrolling behavior, and that discrepancy/overlap is something that needs to be resolved as part of any design/implementation. 

Accomodating zoom in/out would be reasonably straightforward for macOS and mobile platforms, because multi-touch "pinch to zoom" has no ambiguity or overlap, and isn't currently being handled. However, I don't have enough experience with scroll wheel handling on Linux/Windows to comment on whether attaching the scroll wheel to the zoom action on canvas would result in unexpected behavior in any ScrollContainer wrapped around the canvas.
